### PR TITLE
[WASI]: Make epoll edge triggered.

### PR DESCRIPTION
**Pull request by MediosZ**
Addressing #1974 
Signed-off-by: Tricster <mediosrity@gmail.com>

### CODE REVIEW

File: include/host/wasi/inode.h [hunk 1]
```diff
@@ -50,24 +47,32 @@ namespace WASI {
...
 
   constexpr FdHolder() noexcept = default;
-  ~FdHolder() noexcept { reset(); }
-  explicit constexpr FdHolder(int Fd) noexcept : Fd(Fd) {}
+  ~FdHolder() noexcept {
+    if (cleanup) {
+      reset();
+    }
+  }
+  explicit constexpr FdHolder(int Fd, bool cleanup = true) noexcept
```
Comment (hydai) [inline include/host/wasi/inode.h:64]: The first character of the variables should be upper cases.

File: include/host/wasi/inode.h [hunk 2]
```diff
@@ -50,24 +47,32 @@ namespace WASI {
...
   constexpr bool ok() const noexcept { return Fd >= 0; }
   void reset() noexcept;
   int release() noexcept { return std::exchange(Fd, -1); }
   void emplace(int NewFd) noexcept {
     reset();
     Fd = NewFd;
   }
+  int getFd() noexcept { return Fd; }
   int Fd = -1;
+  bool cleanup = true;
```
Comment (hydai) [inline include/host/wasi/inode.h:75]: Ditto.

File: include/host/wasi/environ.h [hunk 3]
```diff
@@ -1155,13 +1168,65 @@ class EVPoller : private VPoller {
...
 inline WasiExpect<EVPoller>
 Environ::pollOneoff(__wasi_size_t NSubscriptions) noexcept {
   return VINode::pollOneoff(NSubscriptions).map([this](VPoller &&P) {
     return EVPoller(std::move(P), *this);
   });
 }
 
+inline WasiExpect<EVEpoller>
+Environ::epollOneoff(__wasi_size_t NSubscriptions) noexcept {
+  auto evepoller =
```
Comment (hydai) [inline include/host/wasi/environ.h:1223]: Ditto.

File: lib/host/wasi/inode-linux.cpp [hunk 4]
```diff
@@ -1592,6 +1699,251 @@ WasiExpect<void> Poller::wait(CallbackType Callback) noexcept {
...
+
+  epoll_event EPollEvent;
+  EPollEvent.events = EPOLLET | EPOLLIN;
+#if defined(EPOLLRDHUP)
+  EPollEvent.events |= EPOLLRDHUP;
+#endif
+  EPollEvent.data.fd = Fd.Fd;
+  // check if fd exists
+  // insert read with fd * 2
+  auto events = EPollEvent.events;
```
Comment (hydai) [inline lib/host/wasi/inode-linux.cpp:1760]: Ditto.

File: lib/host/wasi/inode-linux.cpp [hunk 5]
```diff
@@ -1592,6 +1699,251 @@ WasiExpect<void> Poller::wait(CallbackType Callback) noexcept {
...
+    return WasiUnexpect(__WASI_ERRNO_NOMEM);
+  }
+  epoll_event EPollEvent;
+  EPollEvent.events = EPOLLET | EPOLLOUT;
+#if defined(EPOLLRDHUP)
+  EPollEvent.events |= EPOLLRDHUP;
+#endif
+  EPollEvent.data.fd = Fd.Fd;
+  // insert write with fd * 2 + 1
+  auto events = EPollEvent.events;
```
Comment (hydai) [inline lib/host/wasi/inode-linux.cpp:1814]: Ditto.

File: lib/host/wasi/inode-linux.cpp [hunk 6]
```diff
@@ -1592,6 +1699,251 @@ WasiExpect<void> Poller::wait(CallbackType Callback) noexcept {
...
+Epoller::wait(CallbackType Callback,
+              std::unordered_map<int, uint32_t> &Registration) noexcept {
+  std::vector<struct epoll_event> EPollEvents;
+  try {
+    EPollEvents.resize(Events.size());
+  } catch (std::bad_alloc &) {
+    return WasiUnexpect(__WASI_ERRNO_NOMEM);
+  }
+  std::vector<int> SavedFds;
+  for (auto pair : Registration) {
```
Comment (hydai) [inline lib/host/wasi/inode-linux.cpp:1858]: Ditto.

File: lib/host/wasi/inode-linux.cpp [hunk 7]
```diff
@@ -1592,6 +1699,251 @@ WasiExpect<void> Poller::wait(CallbackType Callback) noexcept {
...
+  }
+
+  std::sort(SavedFds.begin(), SavedFds.end());
+  std::sort(IncomingFds.begin(), IncomingFds.end());
+
+  std::vector<int> Difference;
+  std::set_difference(SavedFds.begin(), SavedFds.end(), IncomingFds.begin(),
+                      IncomingFds.end(), std::back_inserter(Difference));
+
+  for (auto fd : Difference) {
```
Comment (hydai) [inline lib/host/wasi/inode-linux.cpp:1876]: Ditto.

**Review by hydai — CHANGES_REQUESTED** [review]

**Review by hydai — APPROVED** [review]

**Comment by q82419** [comment]
Close #1974

**Pull request merged**